lib/block_view_lockups.go

```diff
@@ -0,0 +1,2178 @@
+	}
+
+	// Validate the transactor as having sufficient DAO Coin or DESO balance for the transaction.
+	var transactorBalanceNanos256 *uint256.Int
+	var prevTransactorBalanceEntry *BalanceEntry
+	var prevCoinEntry *CoinEntry
+	if profilePKID.IsZeroPKID() {
+		// NOTE: When spending balances, we need to check for immature block rewards. Since we don't have
+		//       the block rewards yet for the current block, we subtract one from the current block height
+		//        when spending balances.
+
+		// Check the DeSo balance of the user.
+		transactorBalanceNanos, err := bav.GetSpendableDeSoBalanceNanosForPublicKey(txn.PublicKey, blockHeight-1)
+		if err != nil {
+			return 0, 0, nil, errors.Wrap(err, "_connectCoinLockup")
+		}
+
+		// Construct a uint256 balance and validate the transactor as having sufficient DeSo.
+		transactorBalanceNanos256, _ = uint256.FromBig(big.NewInt(0).SetUint64(transactorBalanceNanos))
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientDeSo, "_connectCoinLockup")
+		}
+
+		// Spend the transactor's DeSo balance.
+		lockupAmount64 := txMeta.LockupAmountBaseUnits.Uint64()
+		newUtxoOp, err := bav._spendBalance(lockupAmount64, txn.PublicKey, blockHeight-1)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(err, "_connectCoinLockup")
+		}
+		utxoOpsForTxn = append(utxoOpsForTxn, newUtxoOp)
+	} else {
+		// Check the BalanceEntry of the user.
+		transactorBalanceEntry, _, _ := bav.GetBalanceEntryForHODLerPubKeyAndCreatorPubKey(
+			txn.PublicKey,
+			txMeta.ProfilePublicKey.ToBytes(),
+			true)
+		if transactorBalanceEntry == nil || transactorBalanceEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinLockupBalanceEntryDoesNotExist, "_connectCoinLockup")
+		}
+
+		// Validate the balance entry as having sufficient funds.
+		transactorBalanceNanos256 = transactorBalanceEntry.BalanceNanos.Clone()
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientCoins, "_connectCoinLockup")
+		}
+
+		// We store the previous transactor balance entry in the event we need to revert the transaction.
+		prevTransactorBalanceEntry = transactorBalanceEntry.Copy()
+
+		// Spend the transactor's DAO coin balance.
+		transactorBalanceEntry.BalanceNanos =
+			*uint256.NewInt().Sub(&transactorBalanceEntry.BalanceNanos, txMeta.LockupAmountBaseUnits)
+		bav._setDAOCoinBalanceEntryMappings(transactorBalanceEntry)
+
+		// Create a copy of the associated CoinEntry in the event we must roll back the transaction.
+		prevCoinEntry = profileEntry.DAOCoinEntry.Copy()
+
+		// Update CoinsInCirculation and NumberOfHolders associated with the DAO coin balance.
+		profileEntry.DAOCoinEntry.CoinsInCirculationNanos = *uint256.NewInt().Sub(
+			&profileEntry.DAOCoinEntry.CoinsInCirculationNanos,
+			txMeta.LockupAmountBaseUnits)
+		if transactorBalanceEntry.BalanceNanos.IsZero() {
```

Comment:
I think you'd be extra safe if you add `&& !prevTransactorBalanceEntry.IsZero()`. This covers an edge case where the balance you fetched was zero. It can't ever happen because you check that you're locking up >0 coins, but still good to be extra safe.